homeassistant/components/sensor/modbus.py

```diff
@@ -55,38 +62,73 @@
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the Modbus sensors."""
     sensors = []
+    data_types = {DATA_TYPE_INT: {1: 'h', 2: 'i', 4: 'q'}}
+    data_types[DATA_TYPE_UINT] = {1: 'H', 2: 'I', 4: 'Q'}
+    data_types[DATA_TYPE_FLOAT] = {1: 'e', 2: 'f', 4: 'd'}
+
     for register in config.get(CONF_REGISTERS):
+        structure = '>i'
+        if register.get(CONF_DATA_TYPE) != DATA_TYPE_CUSTOM:
+            try:
+                structure = '>%c' % (data_types[
+                    register.get(CONF_DATA_TYPE)][register.get(CONF_COUNT)])
+            except KeyError:
+                _LOGGER.error("Unable to detect data type for %s sensor, "
```

Comment:
With the current config validation schema, the possible key error will not be for the data type but for the count. Config validation will make sure that only valid data types are passed in the config. Either change the log message, or if possible you could limit the valid values for `CONF_COUNT` in the schema to make the key error impossible all together.

Reply:
Hi! My english is far away to be perfect, but message seems to be correct. I have to detect type of data stored in modbus register. To do this I have to check count. If I have CONF_DATA_TYPE int I have to un unpack as short if cont is 1, int if count is 2, long if count is 4. If count is 3 I don't know what is the data type. 

I could duplicate the configuration and create a schema with fixed list of CONF_COUNT if type is int, uint or float. Free value of CONF_COUNT When type is custom.

Another soluton is to define 9 fixed data type and use CONF_COUNT only When data type is custom.

What's better?

Reply:
This solution breaks currents configurations, the method in PR keep they working.

Reply:
Yeah, let's keep it as it is then.